Swift WebBackForwardList (off by default)

adetaylor opened the pull request:
<pre>
Swift WebBackForwardList (off by default)
https://bugs.webkit.org/show_bug.cgi?id=305483
rdar://159589815

Reviewed by NOBODY (OOPS!).

This introduces a Swift version of the existing BackForwardList CoreIPC message
handler.

RATIONALE:

WebKit wishes to use Swift to provide much stronger memory safety in its most
sensitive attack surface: CoreIPC message handlers. This PR prototypes how
we can convert a single, simple, message handler to Swift. In itself, converting
this message handler provides negligible security benefit (in fact, rewrites
usually introduce extra bugs) but having a message handler in Swift paves the
way for future work - in particular, it ensures that any future new subsystems
can be written in Swift rather than C++.

BUILDING:

By default this change has no effect. If the LocalOverride option:
  ENABLE_BACK_FORWARD_LIST_SWIFT=ENABLE_BACK_FORWARD_LIST_SWIFT
is given, then a Swift version of the WebBackForwardList will be used. It is
intended to be functionally the same as the C++ version. A recent SDK will
need to be used.

CODE STRUCTURE:

This code is intended to be almost line-for-line the same as the C++ version,
to ease review, which means it is MORE COMPLEX than a typical Swift version.
For instance, an idiomatic Swift version would use an enum to ensure that
the different data members of the BackForwardList cannot become out of sync.

Swift types can't yet inherit from C++ types so when we're in Swift mode,
we still retain a C++ class to inherit from API::Object.

This PR necessarily includes various #ifdefs throughout the codebase to deal
with the Swift vs C++ version. When/if the Swift version becomes the sole
version, they can be removed. We considered allowing Ref<SwiftType> to
minimize the number of places such ifdefs were required, but consensus
was that this just makes things more confusing long-term, and it's better
to stick with ifdefs for now.

FUTURE WORK:

* If this becomes the default or sole implementation, we may choose to
  restructure it to be more idiomatic Swift (moving some runtime
  branches to be compile-time enforced, as noted above)
* The wrappers for WTF logging and IPC
  MESSAGE_CHECKs are temporary - radars note the tasks to build better
  versions.
* Site isolation may change the C++ version of this code, and those changes
  will need to be reflected in the Swift version.
* This code still produces some compiler warnings, which will be resolved
  by adding annotations to further WTF and WebKit C++ types (e.g. WeakRef).
  These will be raised separately.
* This contains two small changes to modulemaps which are necessary for this
  code to build; the need for them is being explored in rdar://166064408
  and either way they should be removed from this PR in due course.

</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca83c42dc3a6db38f22d89f55c36c39500bda992

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143297 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/15772 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/151973 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/96518 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/8f90d1e6-1121-4e5d-b339-f317a70e025f) 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/16432 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/15853 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/151973 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/96518 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/366a4e57-2f2e-49db-aa68-d6ec09c59272) 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/146246 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/16432 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/151973 "Failed to checkout and rebase branch from PR 55393") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/14bcd308-2206-4d17-8626-b10e5f6b48c5) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/16432 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 wpe-libwebrtc](https://ews-build.webkit.org/#/builders/172/builds/1971 "Failed to checkout and rebase branch from PR 55393") | | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/135296 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/16432 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/154284 "Failed to checkout and rebase branch from PR 55393") | | 
| [❌ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/4109 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/15816 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/154284 "Failed to checkout and rebase branch from PR 55393") | | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/15853 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/15853 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/154284 "Failed to checkout and rebase branch from PR 55393") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/71211 "Failed to checkout and rebase branch from PR 55393") | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/15441 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/168/builds/7073 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/174594 "Failed to checkout and rebase branch from PR 55393") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/15175 "Failed to checkout and rebase branch from PR 55393") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/79161 "Failed to checkout and rebase branch from PR 55393") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/174594 "Failed to checkout and rebase branch from PR 55393") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/15386 "Failed to checkout and rebase branch from PR 55393") | | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/15237 "Failed to checkout and rebase branch from PR 55393") | | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -915,15 +919,25 @@ void WebAutomationSession::traverseHistoryInBrowsingContext(const Inspector::Pro
 
+#if ENABLE(BACKFORWARDLIST_SWIFT)
+    WebBackForwardList backForwardList = page->backForwardList();
+    unsigned backCount = backForwardList.backListCount();
```
Only the top item here is different between the ifdefs

adetaylor replied:
Unfortunately not true - the second and third lines have `.` vs `->`

adetaylor replied:
([internal chat thread context](https://a1391192.slack.com/archives/GQUHTL52R/p1767889125809929) for why these ifdefs are here)

adetaylor commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+    }
+}
+
+extension WTF.String {
```
Comment from @rr-codes :

> you should probably move the WTF.String extensions into their own file like WTFString+Extras.swift, and then it can be like
```
import Foundation
import wtf.Core.text.WTFString

extension WTF.String : LosslessStringConvertible {
    public init(_ string: Swift.String) {
        let ns = string as NSString
        let cf: CFString = ns as CFString
        self = WTF.String(cf)
    }

    public var description: Swift.String {
        unsafe createNSString().get()
    }
}

extension WTF.String : ExpressibleByStringLiteral {
    public init(stringLiteral: Swift.String) {
        self.init(stringLiteral)
    }
}

extension WTF.String : ExpressibleByNilLiteral {
    public init(nilLiteral: ()) {
        self = WTF.nullString().pointee
    }
}
```

rr-codes replied:
(might also be nice if this is done in a separate PR to reduce the size of this one, since this part can be done standalone)

adetaylor replied:
Agreed, if I end up making "production quality" wrappers for any WTF types (instead of just the bare minimum required for the BackForwardList) then I'll raise separate PRs.
I'm not exactly sure how best to structure the `ENABLE_` options for them but I'll think about it.

adetaylor replied:
Resolved in https://github.com/WebKit/WebKit/pull/57287 except putting them in their own file, which I'll discuss over there.

rr-codes commented on Source/JavaScriptCore/JavaScriptCore_Private.modulemap:
```diff
@@ -122,8 +122,8 @@ framework module JavaScriptCore_Private [system] {
-        explicit module * {
-            export *
-        }
+        // Module norms would expose nested modules here, but this causes some Swift issues -
```
these and the WebCore module map changes shouldn't be needed (proof of that is https://github.com/WebKit/WebKit/pull/56806 where I'm using Swift-Cxx interop fine)

adetaylor replied:
The changes in these module files are also responsible for the EWS visionOS failures.

I'm working on this class of things in #56712, but haven't succeeded yet…

adetaylor replied:
One of your recent modularisation changes has removed the need for these adjustments! Removing them from the PR.

rr-codes commented on Source/WebKit/Modules/Internal/WebKitInternalCxx.h:
```diff
@@ -25,5 +25,37 @@
 
+#pragma once
+
+#import "APIArray.h"
```
these should all be made separate sub modules in the webkit internal module map

adetaylor replied:
Per [comment here](https://github.com/WebKit/WebKit/pull/57676/changes#r2753786043) I'm not 100% sold that this is correct for non-Apple platforms.

rr-codes commented on Source/WebKit/Shared/WebBackForwardListItem.h:
```diff
@@ -65,6 +66,10 @@ class WebBackForwardListItem : public API::ObjectImpl<API::Object::Type::BackFor
     const String& url() const;
+#ifdef __swift__
+    // Workaround for rdar://162695942
+    String urlCopy() const { return url(); }
```
if the regular `url()` function is technically unsafe, why not just replace it with this one entirely?

adetaylor replied:
Interesting perspective, I will discuss.

adetaylor replied:
Conclusion from chat discussions seems to be that functions returning `String&` are OK and Swift is making the wrong decision here, so I think this workaround is correct.

rr-codes commented on Source/WebKit/UIProcess/API/C/WKAPICast.h:
```diff
@@ -85,7 +85,12 @@ class QueryPermissionResultCallback;
 class SpeechRecognitionPermissionCallback;
 class UserMediaPermissionCheckProxy;
 class UserMediaPermissionRequestProxy;
+#if ENABLE(BACKFORWARDLIST_SWIFT)
```
```suggestion
#if ENABLE(BACK_FORWARD_LIST_SWIFT)
```

very minor personal suggestion but I find this more readable and consistent (since back and forward and list are separate words and should probably be delimited)

adetaylor replied:
Done

rr-codes commented on Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
```diff
@@ -26,6 +26,10 @@
 #import "config.h"
 #import "WKWebViewMac.h"
 
+#if ENABLE(BACKFORWARDLIST_SWIFT)
```
why is this import needed in this file?

adetaylor replied:
I think I meant to mention that in the commit message but forgot. V good question.

Any C++ file which needs to call methods on a Swift type will need to do this. The reasons are:
* obviously, `WebKit-Swift.h` contains the definitions of the classes
* as a blanket rule, we can't allow headers to include `WebKit-Swift.h` because of the reasons explained in rdar://162862737 (risk of circular dependencies). So the actual implementation C++ file needs to include this.

Not ideal - if you have any cunning plans let me know

rr-codes replied:
but WKWebViewMac.mm doesn’t need to call methods on a Swift type?

adetaylor replied:
It does e.g. line 1188 ` return !!_page->backForwardList().backItem();`

rr-codes commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -915,15 +919,25 @@ void WebAutomationSession::traverseHistoryInBrowsingContext(const Inspector::Pro
 
-    RefPtr targetItem = backForwardList->itemAtIndex(targetIndex);
+#if ENABLE(BACKFORWARDLIST_SWIFT)
+    RefPtr targetItem = RefPtr(backForwardList.itemAtIndex(targetIndex));
```
```suggestion
    RefPtr targetItem = backForwardList.itemAtIndex(targetIndex);
```

can it just be this instead?

adetaylor replied:
Probably yes, but it doesn't zap the `ifdef` because it's `.` vs `->`. See also https://a1391192.slack.com/archives/GQUHTL52R/p1767889125809929 for the decision to do this

adetaylor replied:
Change made, though I think it makes it worse because now the two sides of the ifdef look different

rr-codes commented on Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
```diff
@@ -25,6 +25,11 @@
 
 #import "config.h"
 #import "ViewGestureController.h"
+#include <optional>
```
I think usually we prefer to be consistent with include vs import

adetaylor replied:
oops thank you, I suspect this was my IDE, bad IDE

adetaylor replied:
Fixed

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -26,6 +26,10 @@
 #import "config.h"
 #import "WebViewImpl.h"
 
+#if ENABLE(BACKFORWARDLIST_SWIFT)
```
ditto about "why is this import needed in this file?"

adetaylor replied:
Same answer as above so I'll resolve this thread

rr-codes commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2025-2026 Apple Inc. All rights reserved.
```
Would you be able to separate this file out into its own PR first? Since it will make it easier to review, reduce the size of this PR, and it’s mostly standalone too

adetaylor replied:
I'm OK with that. I asked Basuke and Sihui the extent to which they wanted this split up for review, but I haven't had a reply yet - I will prod them.

adetaylor replied:
Prodding completed - I will split this up over the coming days

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.cpp:
```diff
@@ -809,4 +819,134 @@ String WebBackForwardList::loggingString()
+    MESSAGE_CHECK_BASE(false, process->connection());
+}
+
+bool isPlatformCocoa()
```
this should just use `os(whatever)` from Swift instead of having this function (reference https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements/#Conditional-Compilation-Block)

adetaylor replied:
That would be great. How much confidence do you have that they will have identical semantics? I'll look into it but interested in your gut feel too!

adetaylor replied:
OK I've done it

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.cpp:
```diff
@@ -809,4 +819,134 @@ String WebBackForwardList::loggingString()
+#endif
+}
+
+bool isPlatformCocoaOrGTK()
```
ditto

adetaylor replied:
Resolving - we can discuss in the other thread - if we do replace with `os(whatever)` I'll replace this bit too

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.cpp:
```diff
@@ -809,4 +819,134 @@ String WebBackForwardList::loggingString()
+#endif
+}
+
+bool isMacMimeoPhotoProject()
```
ditto

adetaylor replied:
Resolving - we can discuss in the other thread - if we do replace with `os(whatever)` I'll replace this bit too

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.cpp:
```diff
@@ -809,4 +819,134 @@ String WebBackForwardList::loggingString()
+    return WebCore::BackForwardItemIdentifier::generate();
+}
+
+void doLog(const char* WTF_NONNULL msg)
```
please use the standard Swift Logger API instead of these semi weird ad-hoc functions

adetaylor replied:
We can't do that until/unless we've exhaustively proven that they have the same semantics (in terms of enablement modes, destinations of logging etc.), which is why I've raised a radar to do that. If there's a view that we must not land this until that analysis is done, fair enough — but I personally would rather treat it as a separate task to ensure we do it thoroughly.

adetaylor replied:
A brief look: the `LOG` macro has these properties:
* it depends upon `LOG_DISABLED`
* it respects whether different WTF log channels are enabled or disabled, both at compile time and runtime by the looks of it
* on Apple platforms it goes into OS log, on Android it goes into the Android platform log, if it's on other platforms it goes to stderr
* if the log channel is marked as needing accumulation, it adds it to an accumulator

I assume some of these are important properties so I'll look into whether we can plug a different back-end into the Swift logger API to match these things.

(our ref rdar://168139823)

adetaylor replied:
I think the best option here is that we create an alternative `WTFLogger` which is API-compatible with the standard Swift `Logger`, and start to use that. In parallel we  request that the standard Swift logger get pluggable backends, in the hopes that one day we'll be able to replace `WTFLogger` with standard Swift `Logger`. WDYT @rr-codes ?

adetaylor replied:
We also discussed whether we can use the `os_log_hook` interception functions to ensure Swift logging complies with the expectations of WebKit logging,

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+typealias BackForwardListState = WebKit.BackForwardListState
+typealias FrameState = WebKit.FrameState
+typealias WebPageProxy = WebKit.WebPageProxy
+typealias WebFrameProxy = WebKit.WebFrameProxy
```
I am _strongly_ against having these type aliases; these types should just be used as they were intentionally written, with namespaces and all (even in WebKit C++ headers, WebCore types have the WebCore:: prefix)

adetaylor replied:
OK I'll zap them, it's fine with me

adetaylor replied:
Zapped

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+
+typealias BackForwardItemIdentifier = WebCore.BackForwardItemIdentifier
+typealias BackForwardFrameItemIdentifier = WebCore.BackForwardFrameItemIdentifier
+typealias FrameIdentifier = WebCore.FrameIdentifier
```
please import the `WebCore_Private` module since this file uses WebCore types (not entirely sure how it’s even working without it atm)

adetaylor replied:
Done

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+
+// Temporary partial WTF logging support from Swift
+// rdar://168139823 is the task of exposing WebKit logging properly in Swift
+private func backForwardLog(msgCreator: () -> String) {
```
ditto about using the standard Logging API instead

adetaylor replied:
I'm going to resolve this thread so we can discuss pros and cons about this in the comment thread further up

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+}
+
+final class WebBackForwardList {
+    static let defaultCapacity = 100
```
can this be private?

adetaylor replied:
Probably yes, will do

adetaylor replied:
Done

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        // to a map within WebBackForwardListItem. Maybe WebBackForwardList should
+        // own that map. This is a pre-existing quirk of the C++ implementation, not a
+        // new Swift thing.
+        guard let page = page.get() else {
```
this is more of a philosophical question, but how does this guarantee that `page` won’t actually become nil at some point later on if the weakptr gets destroyed?

adetaylor replied:
Because `WeakPtr::get` semantics are `SWIFT_RETURNS_UNRETAINED` (I am planning to raise a PR to make that explicit but I believe it's the default anyway) so Swift will increment the reference count of the returned `page`, until it doesn't need it any more. Right?

adetaylor replied:
@rr-codes let me know if that SGTY and if so I'll resolve this thread

adetaylor replied:
Discussed. This is good

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        currentIndex = nil
+    }
+
+    private func assertStateOk() {
```
I think this should be named named like ```assertValidIndex` or something

adetaylor replied:
Done

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        self.messageForwarder!
+    }
+
+    func itemForID(identifier: BackForwardItemIdentifier) -> WebBackForwardListItem? {
```
```suggestion
    func item(id: BackForwardItemIdentifier) -> WebBackForwardListItem? {
```

adetaylor replied:
I am aiming to match the C++ function name to avoid too many ifdefs across the rest of the world which may call into the Swift or C++ versions

adetaylor replied:
Gonna resolve this but reopen if you disagree

rr-codes commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            return nil
+        }
+
+        let item = WebBackForwardListItem.itemForID(identifier)
```
wait how does this work since itemForID isn’t a static function?

adetaylor replied:
Yes it is - it's a static function on `WebBackForwardList*Item*`

adetaylor commented on Source/WebKit/UIProcess/API/C/WKAPICast.h:
```diff
@@ -85,7 +85,12 @@ class QueryPermissionResultCallback;
+class WebBackForwardListAPIImpl;
+#else
 class WebBackForwardList;
+using WebBackForwardListAPIImpl = WebBackForwardList;
```
Comment from another reviewer offline: we should try to revisit the ownership structure here:
> Normally ClassImpl class is owned by Class to provide implementation
> I think API::Object could hold strong ref to the Impl class, and the Impl might forward that to the Swift class (assuming the Swift class is the only owner of Impl).

adetaylor replied:
I succeeded here. The C++ `WebBackForwardListAPIImpl` is now owned by the Swift `WebBackForwardList`. They hold strong refs to one another and the reference cycle is broken on destruction of the `WebPageProxy`.

aprotyas commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1002 @@
+        // If we're on Google and navigate back, we don't want to skip anything and load Yahoo#a.
+        // However, if we're on Yahoo and navigate forward, we do want to skip items and end up on Google#b.
+        // swift-format-ignore: NeverForceUnwrap
+        if direction == Direction.backward && currentItem()!.wasCreatedByJSWithoutUserInteraction() {
```
there seems to be a small logic inversion here when compared to `WebBackForwardList.cpp`:

https://github.com/WebKit/WebKit/blob/3b652a7e534249327c043126efa94def82510bbf/Source/WebKit/UIProcess/WebBackForwardList.cpp#L552

adetaylor replied:
Thank you!

adetaylor commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2143,7 +2157,7 @@ RefPtr<API::Navigation> WebPageProxy::loadRequest(WebCore::ResourceRequest&& req
         launchProcess(Site { request.url() }, ProcessLaunchReason::InitialProcess);
 
-    Ref navigation = m_navigationState->createLoadRequestNavigation(legacyMainFrameProcess().coreProcessIdentifier(), ResourceRequest(request), m_backForwardList->protectedCurrentItem());
+    Ref navigation = m_navigationState->createLoadRequestNavigation(legacyMainFrameProcess().coreProcessIdentifier(), ResourceRequest(request), m_backForwardList->currentItem());
```
I should look into switching these back to `protectedCurrentItem`

adetaylor replied:
Fixed

adetaylor commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8397,7 +8423,11 @@ void WebPageProxy::decidePolicyForNavigationAction(Ref<WebProcessProxy>&& proces
 
     if (!navigation) {
+#if ENABLE(BACK_FORWARD_LIST_SWIFT)
+        auto& backForwardList = m_backForwardList;
```
I should try to move away from `auto` here

adetaylor replied:
Fixed

adetaylor commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2411,7 +2428,7 @@ RefPtr<API::Navigation> WebPageProxy::loadSimulatedRequest(WebCore::ResourceRequ
         launchProcess(Site { simulatedRequest.url() }, ProcessLaunchReason::InitialProcess);
 
-    Ref navigation = m_navigationState->createSimulatedLoadWithDataNavigation(legacyMainFrameProcess().coreProcessIdentifier(), ResourceRequest(simulatedRequest), makeUnique<API::SubstituteData>(Vector(data->span()), ResourceResponse(simulatedResponse), WebCore::SubstituteData::SessionHistoryVisibility::Visible), m_backForwardList->protectedCurrentItem());
+    Ref navigation = m_navigationState->createSimulatedLoadWithDataNavigation(legacyMainFrameProcess().coreProcessIdentifier(), ResourceRequest(simulatedRequest), makeUnique<API::SubstituteData>(Vector(data->span()), ResourceResponse(simulatedResponse), WebCore::SubstituteData::SessionHistoryVisibility::Visible), m_backForwardList->currentItem());
```
Another case where we should go back to `protectedCurrentItem`

adetaylor commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2663,11 +2680,15 @@ RefPtr<API::Navigation> WebPageProxy::goToBackForwardItem(WebBackForwardListFram
 
     Ref process = m_legacyMainFrameProcess;
-    Ref navigation = m_navigationState->createBackForwardNavigation(process->coreProcessIdentifier(), frameItem, m_backForwardList->protectedCurrentItem(), frameLoadType);
+    Ref navigation = m_navigationState->createBackForwardNavigation(process->coreProcessIdentifier(), frameItem, m_backForwardList->currentItem(), frameLoadType);
```
Another case where we should go back to `protectedCurrentItem`

adetaylor committed:
Swift WebBackForwardList (off by default) - API functions
https://bugs.webkit.org/show_bug.cgi?id=306877
rdar://169533751

Reviewed by NOBODY (OOPS!).

We're introducing a Swift version of the Back Forward List. For the general
design and rationale, see
https://github.com/WebKit/WebKit/pull/55393
It will be landed as a series of separate commits in order to ease code review.

This final commit fills out the remaining function bodies in
WebBackForwardList.swift. While the previous commit added the IPC message handlers,
this one fills out all the rest - primarily APIs called by the rest of WebKit and
layers further up the stack.

Even with this final commit, everything remains turned off unless it's enabled
using LocalOverrides.xcconfig to turn on ENABLE_BACK_FORWARD_LIST_SWIFT

szewai commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1025 @@
+        assert(page.get().map { !$0.hasRunningProcess() } ?? (currentIndex == nil))
+    }
+
+    func webPageProxyDestroyed() {
```
Is there a way we could avoid the ref cycle?
Like:
```
WebPageProxy
 ↓ (Ref - strong)
C++ WebBackForwardListAPIImpl
 ↓ (strong)
Swift WebBackForwardList
 ↓ (C++ WeakPtr - weak)
WebPageProxy
```
I asked to see if we could have Swift object hold strong ref to the C++ object and the intention is that `WebPageProxy` should not hold two members for backforward list (which actually refer to the same list). Checking the implementation and thinking about the ownership, it might be simpler to have `WebPageProxy` still hold strong ref to C++ object and have Swift object be the "implementation".

adetaylor replied:
OK - that should work.

In which case, let's talk about naming! Do you think I should rename:
(C++) `WebBackForwardListAPIImpl` -> `WebBackForwardList`
(Swift) `WebBackForwardList` -> `WebBackForwardListImpl`?

Or something else?

adetaylor replied:
Ah, I've realised that renaming the Swift `WebBackForwardList` can't (easily) be done.

The IPC message receiver needs to be called `WebBackForwardList` whether it's in C++ or Swift, because the IPC code generator (quite reasonably) assumes it matches `WebBackForwardList.messages.in`. Depending on the value of `ENABLE_BACK_FORWARD_LIST_SWIFT`, the implementation of this message receiver may be in C++ or Swift.

So how's about:
(C++) `WebBackForwardListAPIImpl` -> `WebBackForwardListAPI`
(Swift) `WebBackForwardList` - remains `WebBackForwardList`

szewai replied:
Right, the message name and MessageReciever class need to match; I think the C++ class might be renamed to `WebBackForwardListWrapper`

adetaylor commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+
+// Conform any WTF::Vector<WTF::Ref<T>> to this protocol to get useful extensions and iterators
+protocol CxxRefVector {
+    associatedtype Element: CxxRef // you only need to specify this in your conformance
```
Discussing with @rr-codes we think there might be an existing protocol within C++ interop which makes C++ iterators conform to Swift iterator norms. See if I can look that up and use that instead.

adetaylor replied:
https://www.swift.org/documentation/cxx-interop/#working-with-c-containers

adetaylor replied:
It might be possible to get rid of lots of this if it works

adetaylor commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+extension API.Array {
+    // Can't be a designated initializer because we can't see private fields.
+    static func create(list: [API.Object]) -> API.RefAPIArray {
+        var vec = API.VectorAPIObject()
```
This probably should be called `API.VectorRefPtrAPIObject` or something

adetaylor replied:
Done

adetaylor commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+    static func create(list: [API.Object]) -> API.RefAPIArray {
+        var vec = API.VectorAPIObject()
+        vec.reserveCapacity(list.count)
+        for item in list {
```
Maybe I can do this with a `map` or something

adetaylor replied:
This doesn't seem easier since we'd need an initialiser for `VectorAPIObject` which takes a sequence. We could use `forEach` but this is nice and clear already. So not intending to change here.

adetaylor commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+    init(array: [Element.Pointee]) {
+        var vec = Self()
+        vec.reserveCapacity(array.count)
+        for item in array {
```
If this is still necessary maybe this can be a map as well

adetaylor replied:
Again no real benefit per the above comment.

adetaylor commented on Source/WebKit/UIProcess/StdlibExtras.swift:
```diff
@@ -0,0 +1,127 @@
+// Iterator for WTF::Vectors of Ref types
+// We do not attempt to conform the Vector itself to Sequence since this would
+// require C++ interop types to be public
+struct CxxRefVectorIterator<Vec: CxxRefVector>: Sequence, IteratorProtocol {
```
Again hoping this might not be necessary

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.cpp:
```diff
@@ -782,4 +785,107 @@ String WebBackForwardList::loggingString()
+
+void doLoadingReleaseLog(const char* WTF_NONNULL msg)
+{
+    RELEASE_LOG(Loading, "%s", msg);
```
We might be able to switch this over to standard Swift logging sooner because this is much simpler than the `LOG` macro. Something like:

```
extension Logger {
#if os(Cocoa)
   static var backForwardList = Logger(subsystem: "BackForwardList")
#else if os(Windows)
  static var backForwardList = /* some custom log function that just forwards to the old macro */
```

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+// unwrapping so that's been retained.
+
+// rdar://164119356 may allow us to automate some of these conformances
+// using SWIFT_CONFORMS_TO_PROTOCOL
```
Maybe update comment if this proves to be impossible

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+private func backForwardLog(msgCreator: () -> String) {
+    // rdar://133777029 likely will allow us to avoid the performance penalty
+    // of creating the string if logging is disabled.
+    let msg = msgCreator()
```
Make the message creator an auto closure

adetaylor replied:
Parameter label should be _ so callers don't have to specify msgCreator

adetaylor replied:
Done

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+    }
+}
+
+private func loadingReleaseLog(msgCreator: () -> String) {
```
Autoclosure

adetaylor replied:
And equivalent change to parameter name

adetaylor replied:
Done

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+
+// Temporary partial MESSAGE_CHECK_BASE support from Swift
+// Idiomatic equivalent represented by rdar://168139740
+private func messageCheck(process: WebKit.RefWebProcessProxy, assertion: () -> Bool) -> Bool {
```
Consider auto closures and stuff here too

adetaylor replied:
Done

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+    messageCheckCompletion(process: process, assertion: assertion, completionHandler: {})
+}
+
+private func messageCheckCompletion(process: WebKit.RefWebProcessProxy, assertion: () -> Bool, completionHandler: () -> Void) -> Bool {
```
And here

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+    var messageForwarder: RefWebBackForwardListMessageForwarder?
+
+    var entries: [WebKit.WebBackForwardListItem] = []
+    var currentIndex: Array.Index?
```
Make it `Int`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        })
+
+        // We should have always started out with an m_page and we should never close the page twice
+        assert(page.__convertToBool())
```
Zap all `__convertToBool`s

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            return nil
+        }
+
+        // We can't use == here due to rdar://162357139
```
See comment on other PR about contentsMatch being replaceable

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            let targetSize = initialCurrentIndex + 1
+            removedItems.reserveCapacity(entries.count - targetSize)
+
+            while entries.count > targetSize {
```
See if we replace this with a `while let` while matching the C++ precisely

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+                removedItems.append(entries.removeLast())
+            }
+
+            while !entries.isEmpty {
```
This probably _can_ be `while let` while matching C++ closely enough

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        } else {
+            // If we have no current item index we should also not have any entries.
+            // (This in future could be proven by using a Swift enum.)
+            assert(entries.isEmpty)
```
`guard entries.isEmpty else assertionFailure` otherwise the rest

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            return
+        }
+
+        let targetIndex = entries.firstIndex(where: { identitiesMatch($0, item) })
```
Idiomatic way is `firstIndex { ... }`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        // If the target item wasn't even in the list, there's nothing else to do.
+        guard var targetIndex else {
+            backForwardLog(msgCreator: {
+                let identifier = String(item.identifier().toString())
```
Explicit string conversion here won't be needed

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            let deltaValue = if delta > 10 { "over10" } else { delta.description }
+            page.logDiagnosticMessage(
+                WebCore.DiagnosticLoggingKeys.backNavigationDeltaKey(),
+                WTF.String(swiftString: deltaValue),
```
Again not needed long term

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            page.logDiagnosticMessage(
+                WebCore.DiagnosticLoggingKeys.backNavigationDeltaKey(),
+                WTF.String(swiftString: deltaValue),
+                WebCore.ShouldSample.No
```
`.No`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        if !shouldKeepCurrentItem {
+            removedItems.append(entries.remove(at: priorCurrentIndex))
+            // swift-format-ignore: NeverForceUnwrap
+            targetIndex = entries.firstIndex(where: { identitiesMatch($0, item) })!
```
Trailing closure

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            return nil
+        }
+
+        guard let currentIndex = currentIndex else {
```
`guard let currentIndex` and same for following function and any others

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        return entries[index + currentIndex]
+    }
+
+    func backListCount() -> Array.Index {
```
Replace with `Int`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        let entriesCopy = entries
+        entries.removeAll()
+        // swift-format-ignore: NeverForceUnwrap
+        page.get()!.didChangeBackForwardList(Optional.none, consuming: WebKit.BackForwardListItemVector(array: entriesCopy))
```
Check that C++ intended to crash if page is invalid

adetaylor replied:
Use `nil`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+    func backForwardListState(filter: WebBackForwardListItemFilter) -> WebKit.BackForwardListState {
+        assertValidIndex()
+
+        var backForwardListState = WebKit.BackForwardListState.init()
```
Zap `.init`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        }
+
+        for (i, entry) in entries.enumerated() {
+            if !filter.pointee(entry) {
```
Could make a Swift extension for the `WTF::RefCountable` with `callAsFunction`

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+
+    private func setBackForwardItemIdentifiers(frameState: WebKit.FrameState, itemID: WebCore.BackForwardItemIdentifier) {
+        frameState.itemID = WebCore.MarkableBackForwardItemIdentifier(itemID)
+        frameState.frameItemID = WebCore.MarkableBackForwardFrameItemIdentifier(generateBackForwardFrameItemIdentifier())
```
See if the weak linkage problem rdar://162193891 is fixed

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            entries.append(item.ptr())
+        }
+
+        currentIndex = Optional(fromCxx: backForwardListState.currentIndex).map({ val in Int(val) })
```
Can be `.map(Int(_:))` ...?

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        page.get()!.backForwardRemovedItem(item.identifier())
+
+        // rdar://168139870 to clean up use of BUILDING_GTK__ here.
+        #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS) || os(visionOS) || BUILDING_GTK__
```
If `os(Darwin)`?

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            return WebKit.RefPtrWebBackForwardListItem()
+        }
+
+        #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS) || os(visionOS)
```
`os(Darwin)`?

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+            }
+        var itemIndex = delta
+        let item = itemAtIndex(index: itemIndex)
+        guard var item = item else {
```
`guard var item` (and check it can't be `let`)

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        // If we're on Google and navigate back, we don't want to skip anything and load Yahoo#a.
+        // However, if we're on Yahoo and navigate forward, we do want to skip items and end up on Google#b.
+        // swift-format-ignore: NeverForceUnwrap
+        if direction == Direction.backward && !currentItem()!.wasCreatedByJSWithoutUserInteraction() {
```
`.backward` and do equivalent everywhere

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        let itemURL = unsafe WTF.URL(navigatedFrameState.ptr().urlString, nil)
+        let itemOriginalURL = unsafe WTF.URL(navigatedFrameState.ptr().originalURLString, nil)
+
+        #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS) || os(visionOS)
```
`os(Darwin)`?

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        addItem(newItem: item)
+    }
+
+    // IPCs from here on
```
Consider grouping them into an extension

adetaylor commented on Source/WebKit/UIProcess/WebBackForwardList.swift:
```diff
@@ -0,0 +1,1003 @@
+        if let page = page.get() {
+            let loadedWebArchive =
+                page.didLoadWebArchive()
+                ? WebKit.LoadedWebArchive.Yes
```
.Yes etc.

adetaylor replied:
Done

adetaylor committed:
A couple of SWIFT_RETURNED_AS_UNRETAINED_BY_DEFAULT

adetaylor committed:
WeakPtr add SWIFT_RETURNS_UNRETAINED

adetaylor committed:
Remove CxxRefVectorIterator

Switch away from using CxxRefVectorIterator

adetaylor committed:
Revert "Remove CxxRefVectorIterator"

This reverts commit 48648461b72585b6060a0fcad2397b13a607d3d5.

adetaylor committed:
-clang-target arm64e-apple-macos26.4

adetaylor committed:
Reinstate equality function
https://bugs.webkit.org/show_bug.cgi?id=307163
rdar://169799223

Reviewed by NOBODY (OOPS!).

Part of the Web Back Forward List Swift code was a func === implementation to
allow comparison of C++ types by identity. This relied on an @_expose(!Cxx)
attribute to avoid exposing this back to C++; that was not supported on some
Swift versions between 6.2.0 and 6.2.3. We therefore had to tempoararily remove
that function to fix a build failure. Reinstate the function, but bump our
Swift conditional compilation guards to 6.2.3 so that it does not attempt to
parse this attribute.
